datafusion/sql/src/select.rs

```diff
@@ -194,7 +198,13 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                 .iter()
                 .map(|expr| rebase_expr(expr, &window_func_exprs, &plan))
                 .collect::<Result<Vec<Expr>>>()?;
-
+            if select.into.is_some() {
+                for expr in select_exprs_post_aggr.iter_mut() {
+                    if let Expr::Column(_) = expr.clone() {
+                        *expr = expr.clone().alias(physical_name(expr)?);
+                    }
+                }
+            }
```

Comment:
I am sorry if my past comments have been confusing. Here is what I was trying to say earlier in https://github.com/apache/arrow-datafusion/pull/6566#discussion_r1221355361:

I ran this command to get some logs (with the extra debug in PR https://github.com/apache/arrow-datafusion/pull/6626):

```shell
RUST_LOG=debug cargo test --test sqllogictests -- ddl 2>&1 | tee /tmp/debug.log
```

Here is the content of debug.log:  [debug.log](https://github.com/apache/arrow-datafusion/files/11712340/debug.log)



From the log, here is the `LogialPlan` that shows the  `WindowAggr` declares it makes a column named `SUM(test_table.c1) ORDER BY [test_table.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW` (yes that whole thing!)

```
    Projection: SUM(test_table.c1) ORDER BY [test_table.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW, test_table.c2, test_table.c3
      WindowAggr: windowExpr=[[SUM(test_table.c1) ORDER BY [test_table.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW]]
        TableScan: test_table projection=[c1, c2, c3]
```


Here is the final `ExecutionPlan`, also showing the same giant column as the declared output name:

```
[2023-06-10T11:43:29Z DEBUG datafusion::physical_plan::planner] Optimized physical plan:
    ProjectionExec: expr=[SUM(test_table.c1) ORDER BY [test_table.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW@3 as SUM(test_table.c1), c2@1 as c2, c3@2 as c3]
      BoundedWindowAggExec: wdw=[SUM(test_table.c1): Ok(Field { name: "SUM(test_table.c1)", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(Float32(NULL)), end_bound: CurrentRow }], mode=[Sorted]
        SortPreservingMergeExec: [c2@1 ASC NULLS LAST]
          SortExec: expr=[c2@1 ASC NULLS LAST]
            MemoryExec: partitions=4, partition_sizes=[1, 0, 0, 0]
```

However, looking at the logs, what the execution plan actually produces a column named `SUM(test_table.c1)`:

```
[2023-06-10T11:43:29Z DEBUG datafusion::datasource::memory] mem schema does not contain batches schema.

Target_schema: Schema { fields: [
  Field { name: "SUM(test_table.c1) ORDER BY [test_table.c2 ASC NULLS LAST] RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT ROW", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} },
  Field { name: "c2", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} },
  Field { name: "c3", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {}
}.


Batches Schema: Schema { fields: [
  Field { name: "SUM(test_table.c1)", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} },
  Field { name: "c2", data_type: Float32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} },
  Field { name: "c3", data_type: Utf8, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }], metadata: {}
}
```

Thus, what I was trying to say earlier is that I think the root of the problem is the mismatch between what the plans say the field name of the output is and what the field name that the WindowExec is actually producing.

So I think we should fix this bug by resolving the mismatch. Either:
1. Update the Logical/Physical plans so the field names of WindowAgg matches what the `BoundedWindowAggExec` actually produces
2. OR Update `BoundedWindowAggExec` to produce the field names declared by the `WindowAggExec

Reply:
Thanks @alamb @berkaysynnada 
Would you mind if I also open a small PR for this?

Reply:
> Would you mind if I also open a small PR for this?

I would be very much appreciative, personally

Reply:
Sorry for the late reply. I have tried both suggestions @alamb:
1) We need to modify 3 parts of the code: 
a- In the `project()` function in select.rs, the final schema will be constructed with a shortened form of the window function. 
```
    fn to_field(&self, input_schema: &DFSchema) -> Result<DFField> {
        match self {
            Expr::Column(c) => Ok(DFField::new(
                c.relation.clone(),
                &c.name,
                self.get_type(input_schema)?,
                self.nullable(input_schema)?,
            )),
            _ => {
                Ok(DFField::new_unqualified(
                    &self.display_name()?,
                    self.get_type(input_schema)?,
                    self.nullable(input_schema)?,
                ))
            }
        }
    }
```
is expanded with that arm:
                
```
          Expr::WindowFunction(WindowFunction { fun, args, .. }) => {
                Ok(DFField::new_unqualified(
                    &vec![create_function_name(&fun.to_string(), false, args)?].join(" "),
                    self.get_type(input_schema)?,
                    self.nullable(input_schema)?,
                ))
            }
```
b- In the `project()` function again, qualified wildcard columns are normalized. However, the column name is in the longer form, and the schema of the plan is in the shorter form. Therefore, we also change the `expr_as_column_expr()` function so that window function expressions are converted to column expressions with the shortened column name format, which can be copied from the schema.
c- `PushDownProjection` rule again creates new column expressions with `display_name() `function (which returns the long format) in the window handling arm. These column names also need to be shortened to satisfy subsequent control.

**My opinion:** Can we directly change the `display_name()` function for window functions such that only function name and arguments are returned? Thus we don't need to change any of what I mentioned above.
```
        Expr::WindowFunction(WindowFunction {
            fun, args, window_frame, partition_by, order_by,
        }) => {
            let mut parts: Vec<String> =
                vec![create_function_name(&fun.to_string(), false, args)?];
            if !partition_by.is_empty() {
                parts.push(format!("PARTITION BY {partition_by:?}"));
            }
            if !order_by.is_empty() {
                parts.push(format!("ORDER BY {order_by:?}"));
            }
            parts.push(format!("{window_frame}"));
            Ok(parts.join(" "))
        }
```
new version:
```
Expr::WindowFunction(WindowFunction {
            fun, args, ..
        }) => {
            let mut parts: Vec<String> =
                vec![create_function_name(&fun.to_string(), false, args)?];
            Ok(parts.join(" "))
        }
```


2) We can only change `create_window_expr()` such that it creates the window name with `display_name()` rather than `physical_name()`, but it makes the plans longer, also lots of test change burden.

Reply:
@berkaysynnada thanks for checking that.

I was also working on that.
Changing `display_name` was the one I started with but in this case other scenarios will fail. When window plan created the DFS schema check name uniqueness from `display_name` not considering aliases. So this query will fail
```
SELECT
  first_value(c9) OVER (PARTITION BY c2 ORDER BY c9) first_c9,
  first_value(c9) OVER (PARTITION BY c2 ORDER BY c9 DESC) first_c9_desc
FROM aggregate_test_100
```

I'm still thinking how to overcome that without breaking changes

Reply:
After exploring all options, I get the impression that this option that @alamb mentioned before:

> Update `BoundedWindowAggExec` to produce the field names declared by the `WindowAggExec`

is the most consistent path forward. It tends to produce overly verbose column names, but all the other options seems to suffer from some kind of a problem. AFAICT this approach is similar to what @berkaysynnada mentioned as [his first solution alternative in the issue description](https://github.com/apache/arrow-datafusion/issues/6492) as well.

@alamb, @comphead: What do you think? Should we move forward with this approach?

Reply:
> @alamb, @comphead: What do you think? Should we move forward with this approach?

Yes, that is what I think we should do. If the overly verbose column (at the output) names are a problem, perhaps we can look into updating the planner to automatically add more reasonable aliases

Reply:
Agree, we need to get back on column naming convention as currently long names are not user friendly and not useful without aliases in nested queries

Reply:
It should also be considered that we cannot support more than one column with the same aliases, if we intend to shorten the names at the output by making realias.

Reply:
Sounds good. We will go forward with that approach and @berkaysynnada will update you guys of the progress.

Reply:
Related issues: https://github.com/apache/arrow-datafusion/issues/6543 and https://github.com/apache/arrow-datafusion/issues/6758